asmopt/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+    });
+}
+
+/// Analyzes each machine and successively removes unnecessary components:
+/// 1. Removes declarations of instructions that are never used.
+/// 2. Removes instances of submachines that are never used, including those that became unused in the previous step.
+/// 3. Removes unused registers.
+fn asm_remove_unused_machine_components(asm_file: &mut AnalysisASMFile) {
+    let constrained_machines: HashSet<_> = collect_constrained_machines(asm_file).collect();
+    for (_, machine) in asm_file.machines_mut() {
+        let submachine_to_decl: HashMap<String, String> = machine
+            .submachines
+            .iter()
+            .map(|sub| (sub.name.clone(), sub.ty.to_string()))
+            .collect();
+
+        let symbols_in_callable: HashSet<String> = machine_callable_body_symbols(machine).collect();
+
+        machine_remove_unused_instructions(machine, &symbols_in_callable);
+        machine_remove_unused_submachines(
+            machine,
+            &symbols_in_callable,
+            &submachine_to_decl,
+            &constrained_machines,
+        );
+        machine_remove_unused_registers(machine, &submachine_to_decl);
+    }
+}
+
+fn machine_remove_unused_registers(
+    machine: &mut Machine,
+    submachine_to_decl: &HashMap<String, String>,
+) {
+    let used_symbols: HashSet<_> = once(PC_REGISTER.to_string())
+        .chain(machine_callable_body_symbols(machine))
+        .chain(machine_in_links(machine, submachine_to_decl))
+        .chain(machine_instructions_symbols(machine))
+        .collect();
+
+    machine
+        .registers
+        .retain(|reg| used_symbols.contains(&reg.name));
+}
+
+fn machine_remove_unused_submachines(
+    machine: &mut Machine,
+    symbols: &HashSet<String>,
+    submachine_to_decl: &HashMap<String, String>,
+    constrained_machines: &HashSet<String>,
+) {
+    let visited_submachines = machine
+        .instructions
+        .iter()
+        .filter(|ins| symbols.contains(&ins.name))
+        .flat_map(|ins| {
+            ins.instruction
+                .links
+                .iter()
+                .filter_map(|link| submachine_to_decl.get(&link.link.instance))
+        })
+        .cloned();
+
+    let mut used_submachines: HashSet<_> = visited_submachines
+        .chain(machine_callable_body_symbols(machine))
+        .chain(machine_in_links(machine, submachine_to_decl))
+        .chain(machine_in_args(machine, submachine_to_decl))
+        .collect();
+
+    used_submachines.extend(constrained_machines.iter().cloned());
+
+    machine
+        .submachines
+        .retain(|sub| used_submachines.contains(&sub.ty.to_string()));
+}
+
+fn machine_remove_unused_instructions(machine: &mut Machine, symbols: &HashSet<String>) {
+    machine
+        .instructions
+        .retain(|ins| symbols.contains(&ins.name));
+}
+
+fn collect_constrained_machines(asm_file: &AnalysisASMFile) -> impl Iterator<Item = String> + '_ {
+    asm_file
```

Comment:
This function searches for constrained machines by looking for machines where columns are defined, but perhaps there is a more precise way?